src/app/loadout-drawer/loadout-apply.ts

```diff
@@ -552,3 +596,174 @@ export function clearItemsOffCharacter(
+ * mods into any equipped armor, not just armor in the loadout - this allows for loadouts that
+ * are *only* mods to be applied to current armor.
+ *
+ * Right now this will try to apply mods if they'll fit, but if they won't it'll blindly remove
+ * all mods on the piece before adding the new ones. We don't yet take into consideration which
+ * mods are already on the items.
+ */
+// TODO: If we fail to move/equip armor, should we still equip mods to it?
+function applyLoadoutMods(
+  storeId: string,
+  /** A list of inventory item hashes for plugs */
+  modHashes: number[]
+): ThunkResult<number[]> {
+  return async (dispatch, getState) => {
+    const defs = d2ManifestSelector(getState())!;
+    const stores = storesSelector(getState());
+    const store = getStore(stores, storeId)!;
+
+    // TODO: find cases where the loadout specified armor for a slot to be equipped and it's not equipped, bail if so
+    const armor = store.items.filter((i) => i.bucket.inArmor && i.equipped);
+    //const loadoutArmor = loadout.items.filter((i) => i.equipped)
+
+    const mods = modHashes.map((h) => defs.InventoryItem.get(h)).filter(isPluggableItem);
+
+    // What mods are already on the equipped armor set?
+    const existingMods = new Set<number>();
+    for (const item of armor) {
+      if (item.sockets) {
+        for (const socket of item.sockets.allSockets) {
+          if (socket.plugged) {
+            existingMods.add(socket.plugged.plugDef.hash);
+          }
+        }
+      }
+    }
+
+    // Early exit - if all the mods are already there, nothing to do
+    if (modHashes.every((h) => existingMods.has(h))) {
+      infoLog('loadout mods', 'all mods are already there, skipping');
+      return modHashes;
+    }
+
+    // TODO: stop if the API doesn't exist
+    // TODO: stop if we can't perform this action (in activity)
+    // TODO: prefer equipping to armor that *is* part of the loadout
+    // TODO: compute assigments should consider which mods are already on the item!
+    const modAssignments = getCheapestModAssignments(armor, mods, defs);
+
+    const successfulMods: number[] = [];
+
+    for (const item of armor) {
+      const modsForItem = modAssignments.itemModAssignments[item.id];
+      if (modsForItem) {
+        successfulMods.push(...(await dispatch(equipMods(item.id, modsForItem))));
+      }
+    }
+
+    // TODO: better mod assignment (for this problem):
+    // Pass 1: Find out which of the mods are already there
+    // Pass 2: Slot specific mods just go right away
+    // Pass 3: Figure out if we can slot mods into existing spaces
+    // Pass 4: Figure out the minimum mods to remove to get the new mods in
+
+    // TODO: notify unassigned mods!
+
+    // Return the mods that were successfully assigned (even if they didn't have to move)
+    return successfulMods;
+  };
+}
+
+/**
+ * Equip the specified mods on the item. Strips off existing mods if needed.
```

Comment:
Right now it strips off *all* mods that aren't part of the assignment, because it means we won't ever have a condition where halfway through equipping mods we go over the energy limit. I think we can do better here pretty easily, I just wanted to start with the easiest.

Reply:
Yeah I think we can go over the energy limit but it would be best to do this until we verify that. If we can go over the energy limit, it will just be a matter of assigning mods to items and default plugs to things without mods

Reply:
I'd be extremely shocked if we can go over the energy limit.

Reply:
A brief test with the current AWA suggests we cannot.